java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptionStateRecorder.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2016-2017, Joyent, Inc. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.joyent.manta.client.multipart;
+
+import com.joyent.manta.client.crypto.EncryptionContext;
+import com.joyent.manta.exception.MantaReflectionException;
+import com.joyent.manta.util.CipherCloner;
+import com.joyent.manta.util.Cloner;
+import com.joyent.manta.util.HmacCloner;
+import com.joyent.manta.util.HmacOutputStream;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.bouncycastle.crypto.macs.HMac;
+import org.bouncycastle.jcajce.io.CipherOutputStream;
+
+import javax.crypto.Cipher;
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+
+/**
+ * Helper class for recording state of encryption and authentication state during in-progress MPU uploads.
+ * Used to enable retry of {@link EncryptedMultipartManager#uploadPart} in case of network failure.
+ * Like the <a href="https://sourcemaking.com/design_patterns/memento">Memento Pattern</a> but with less respect
+ * for encapsulation.
+ */
+public class EncryptionStateRecorder {
```

Comment:
This class should have more comments on why we would even need a encryption state recorder.